File: matiasmasso/Mat.NET/pwa.matiasmasso.es/Web2/Helpers/PdfBase.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using C1.C1Pdf;
using System.Drawing;
using Microsoft.AspNetCore.Mvc;
using System.Runtime.Versioning;

namespace Web.Helpers
{
    [SupportedOSPlatform("windows")]
    public class PdfBase
    {
        public C1PdfDocument Pdf { get; set; }
        public bool Landscape { get; set; }
        public Font Font { get; set; }
        public Brush Brush { get; set; }
        public Pen Pen { get; set; }
        public int X { get; set; }
        public int Y { get; set; }
        public int marginTop { get; set; }
        public int marginLeft { get; set; }
        public int marginRight { get; set; }
        public int marginBottom { get; set; }

        public List<Exception> exs { get; set; } = new List<Exception>();

        protected double PxToMm = 25.4 / 100;
        protected double MmToPx ;



        public PdfBase(System.Drawing.Printing.PaperKind paperKind = System.Drawing.Printing.PaperKind.A4, bool landscape = false)
        {
            MmToPx = 1 / PxToMm;
            Pdf = new C1PdfDocument(paperKind, landscape);
            Font = new Font("Helvetica", 8, FontStyle.Regular);
            Brush = new SolidBrush(System.Drawing.Color.Black);
            Pen = new Pen(Brush, (float)0.5);
        }

        public void SetMargins(int top = 0, int left = 0, int right = 0, int bottom = 0)
        {
            this.marginTop = top;
            this.marginLeft = left;
            this.marginRight = right;
            this.marginBottom = bottom;
        }

        public int Top()
        {
            return (int)(this.Pdf.PageRectangle.Top + this.marginTop);
        }

        public int Right()
        {
            var retval = this.Pdf.PageRectangle.Right - this.marginRight;
            return (int)retval;
        }

        public int Bottom()
        {
            return (int)(this.Pdf.PageRectangle.Bottom - this.Bottom());
        }

        public int Left()
        {
            var retval = this.Pdf.PageRectangle.Left + this.marginLeft;
            return (int)retval;
        }

        public void NewPage()
        {
            this.Pdf.NewPage();
        }

        public byte[] Stream()
        {
            System.IO.MemoryStream oMemoryStream = new System.IO.MemoryStream();
            Pdf.Save(oMemoryStream);

            byte[] retval = oMemoryStream.ToArray();
            return retval;
        }

        public void DrawStringLine(string sText)
        {
            DrawString(sText);
            Y += Font.Height;
        }

        public void DrawString(string sText, Font? font = default, Brush? brush = default, System.Drawing.RectangleF rc = default, StringFormat? sF = default)
        {
            if (font == null)
                font = Font;
            if (brush == null)
                brush = Brush;
            if (sF == null)
                sF = new StringFormat();
            if (!string.IsNullOrEmpty(sText ))
            {
                System.Drawing.PointF pt;
                if (rc == default )
                    pt = new System.Drawing.PointF(marginLeft + X, marginTop + Y);
                else
                    pt = new System.Drawing.PointF(rc.Left, rc.Top);
                Pdf.DrawString(sText, font, brush, pt, sF);
            }
        }

        public void DrawCenteredString(string sText)
        {
            if (!string.IsNullOrEmpty(sText))
            {
                int iWidth = (int)this.Pdf.MeasureString(sText, Font).Width;
                System.Drawing.PointF pt = new System.Drawing.PointF((float)(marginLeft + X - iWidth / (double)2), marginTop + Y);
                Pdf.DrawString(sText, Font, Brush, pt);
            }
        }

        public void DrawImage(System.Drawing.Image oImage, System.Drawing.RectangleF rc)
        {
            this.Pdf.DrawImage(oImage, rc);
        }

        public void DrawRectangle(int iX, int iY, float iWidth = -1, float iHeight = -1)
        {
            if (iWidth == -1)
                iWidth = this.Pdf.PageRectangle.Width - iX - this.marginLeft - this.marginRight;
            if (iHeight == -1)
                iHeight = this.Pdf.PageRectangle.Height - iY - this.marginTop - this.marginBottom;
            this.Pdf.DrawRectangle(Pen, marginLeft + iX, marginTop + iY, iWidth, iHeight);
        }

        public void DrawPageRectangle()
        {
            System.Drawing.RectangleF rc = Pdf.PageRectangle;
            // rc.Inflate(-_marginLeft - _marginRight, -_marginTop - _marginBottom)
            rc.Inflate(-marginLeft, -marginTop);
            this.Pdf.DrawRectangle(Pen, rc);
        }


        public System.Drawing.Rectangle MarginsRectangle()
        {
            var retval = new System.Drawing.Rectangle(this.marginLeft, this.marginTop, (int)(this.Pdf.PageRectangle.Width - this.marginLeft - this.marginRight), (int)(this.Pdf.PageRectangle.Height - this.marginTop - this.marginBottom));
            return retval;
        }

        public int MeasureStringWidth(string s)
        {
            int retval = (int)Pdf.MeasureString(s, Font).Width;
            return retval;
        }

        public Font GetAdjustedFont(string GraphicString, int ContainerWidth, int MaxFontSize, int MinFontSize, bool SmallestOnFail = false)
        {
            Font? testFont = default;
            for (int AdjustedSize = MaxFontSize; AdjustedSize >= MinFontSize; AdjustedSize += -1)
            {
                testFont = new Font(Font.Name, AdjustedSize, Font.Style);
                System.Drawing.SizeF AdjustedSizeNew = Pdf.MeasureString(GraphicString, testFont);
                if (ContainerWidth > Convert.ToInt32(AdjustedSizeNew.Width))
                    return testFont;
            }

            if (SmallestOnFail)
                return testFont;
            else
                return Font;
        }
    }
}

File: matiasmasso/Mat.NET/LegacyHelper/Pdf/_PdfBase.vb
Imports C1.C1Pdf
Public Class _PdfBase
    Property Pdf As C1PdfDocument
    Property Landscape As Boolean
    Property Font As Font
    Property Brush As Brush
    Property Pen As Pen
    Property X As Integer
    Property Y As Integer
    Property marginTop As Integer
    Property marginLeft As Integer
    Property marginRight As Integer
    Property marginBottom As Integer

    Property exs As New List(Of Exception)

    Protected PxToMm As Decimal = 25.4 / 100
    Protected MmToPx As Decimal = 1 / PxToMm



    Public Sub New(Optional paperKind As Printing.PaperKind = Printing.PaperKind.A4, Optional landscape As Boolean = False)
        _Pdf = New C1PdfDocument(paperKind, landscape)
        _Font = New Font("Helvetica", 8, FontStyle.Regular)
        _Brush = New SolidBrush(Color.Black)
        _Pen = New Pen(_Brush, 0.5)
    End Sub

    Public Sub SetMargins(Optional top As Integer = 0, Optional left As Integer = 0, Optional right As Integer = 0, Optional bottom As Integer = 0)
        Me.marginTop = top
        Me.marginLeft = left
        Me.marginRight = right
        Me.marginBottom = bottom
    End Sub

    Public Function Top() As Integer
        Return Me.Pdf.PageRectangle.Top + Me.marginTop
    End Function

    Public Function Right() As Integer
        Dim retval = Me.Pdf.PageRectangle.Right - Me.marginRight
        Return retval
    End Function

    Public Function Bottom() As Integer
        Return Me.Pdf.PageRectangle.Bottom - Me.Bottom
    End Function

    Public Function Left() As Integer
        Dim retval = Me.Pdf.PageRectangle.Left + Me.marginLeft
        Return retval
    End Function

    Public Sub NewPage()
        Me.Pdf.NewPage()
    End Sub

    Public Function Stream() As Byte()
        Dim oMemoryStream As New IO.MemoryStream
        _Pdf.Save(oMemoryStream)

        Dim retval As Byte() = oMemoryStream.ToArray
        Return retval
    End Function

    Public Sub DrawStringLine(sText As String)
        DrawString(sText)
        Y += _Font.Height
    End Sub

    Public Sub DrawString(sText As String, Optional font As Font = Nothing, Optional brush As Brush = Nothing, Optional rc As RectangleF = Nothing, Optional sF As StringFormat = Nothing)
        If font Is Nothing Then font = _Font
        If brush Is Nothing Then brush = _Brush
        If sF Is Nothing Then sF = New StringFormat
        If sText > "" Then
            Dim pt As PointF
            If rc = Nothing Then
                pt = New PointF(_marginLeft + X, _marginTop + Y)
            Else
                pt = New PointF(rc.Left, rc.Top)
            End If
            _Pdf.DrawString(sText, font, brush, pt, sF)
        End If
    End Sub

    Public Sub DrawCenteredString(sText As String)
        If sText > "" Then
            Dim iWidth As Integer = Me.Pdf.MeasureString(sText, _Font).Width
            Dim pt As New PointF(_marginLeft + X - iWidth / 2, _marginTop + Y)
            _Pdf.DrawString(sText, _Font, _Brush, pt)
        End If
    End Sub

    Public Sub DrawImage(oImage As Image, rc As RectangleF)
        Me.Pdf.DrawImage(oImage, rc)
    End Sub

    Public Sub DrawRectangle(iX As Integer, iY As Integer, Optional iWidth As Single = -1, Optional iHeight As Single = -1)
        If iWidth = -1 Then iWidth = Me.Pdf.PageRectangle.Width - iX - Me.marginLeft - Me.marginRight
        If iHeight = -1 Then iHeight = Me.Pdf.PageRectangle.Height - iY - Me.marginTop - Me.marginBottom
        Me.Pdf.DrawRectangle(_Pen, _marginLeft + iX, _marginTop + iY, iWidth, iHeight)
    End Sub

    Public Sub DrawPageRectangle()
        Dim rc As RectangleF = _Pdf.PageRectangle
        'rc.Inflate(-_marginLeft - _marginRight, -_marginTop - _marginBottom)
        rc.Inflate(-_marginLeft, -_marginTop)
        Me.Pdf.DrawRectangle(_Pen, rc)
    End Sub


    Public Function MarginsRectangle() As Rectangle
        Dim retval As New Rectangle(Me.marginLeft, Me.marginTop, Me.Pdf.PageRectangle.Width - Me.marginLeft - Me.marginRight, Me.Pdf.PageRectangle.Height - Me.marginTop - Me.marginBottom)
        Return retval
    End Function

    Public Function MeasureStringWidth(s) As Integer
        Dim retval As Integer = _Pdf.MeasureString(s, _Font).Width
        Return retval
    End Function

    Public Function GetAdjustedFont(ByVal GraphicString As String, ByVal ContainerWidth As Integer, ByVal MaxFontSize As Integer, ByVal MinFontSize As Integer, Optional ByVal SmallestOnFail As Boolean = False) As Font
        Dim testFont As Font = Nothing
        For AdjustedSize As Integer = MaxFontSize To MinFontSize Step -1
            testFont = New Font(_Font.Name, AdjustedSize, _Font.Style)
            Dim AdjustedSizeNew As SizeF = _Pdf.MeasureString(GraphicString, testFont)
            If ContainerWidth > Convert.ToInt32(AdjustedSizeNew.Width) Then
                Return testFont
            End If
        Next

        If SmallestOnFail Then
            Return testFont
        Else
            Return _Font
        End If
    End Function
End Class
